src/io/mod.rs

```diff
@@ -0,0 +1,330 @@
+                {
+                    partition.push(file);
+                    curr_part_size = new_part_size;
+                } else if size > avg_partition_size as u64 {
+                    partitions.push(partition);
+                    partitions.push(vec![file]);
+                    partition = vec![];
+                    curr_part_size = 0;
+                } else {
+                    partitions.push(partition);
+                    partition = vec![file];
+                    curr_part_size = size;
+                }
+            }
+            partitions.push(partition);
+            partitions
+        }
+    }
+
+    impl DistributedReader for LocalFsReader {
+        fn get_reading_handler(self) -> Box<dyn Read> {
+            let LocalFsReader { mut files, .. } = self;
+
+            Box::new(LocalExecutorFsReader {
+                files: files.pop().expect("executor failed to fetch files to read"),
+                open_file: None,
+            })
+        }
+    }
+
+    impl Chunkable<Self> for LocalFsReader {
+        fn slice_with_set_parts(self, parts: usize) -> Vec<Arc<Vec<LocalFsReader>>> {
+            let mut files_by_part = self.load_local_files().expect("failed to load local files");
+            // for each chunk we create a new loader to be run in parallel
+            let mut loaders = Vec::with_capacity(files_by_part.len());
+            for chunk in files_by_part {
+                // a bit hacky, but the only necessary attribute in the next phase is `files`
+                let partial_loader = LocalFsReader {
+                    path: PathBuf::new(),
+                    is_single_file: false,
+                    filter_ext: None,
+                    expect_dir: true,
+                    files: vec![chunk],
+                    executor_partitions: 1,
+                };
+                loaders.push(Arc::new(vec![partial_loader]));
+            }
+            loaders
+        }
+    }
+
+    struct LocalExecutorFsReader {
+        files: Vec<PathBuf>,
+        open_file: Option<BufReader<fs::File>>,
+    }
+
+    impl Read for LocalExecutorFsReader {
+        // TODO: optimally read should be async so the executors can do other work meanwhile,
+        // and shouldn't run too many parallel reads per worker node as it's wasteful.
+        fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+            if let Some(open_reader) = self.open_file.as_mut() {
+                open_reader.read_exact(buf)?;
+                Ok(buf.len())
+            } else if let Some(mut file) = self.files.pop() {
+                let file = fs::File::open(file)?;
+                let mut new_reader = BufReader::new(file);
+                new_reader.read_exact(buf)?;
+                self.open_file = Some(new_reader);
+                Ok(buf.len())
+            } else {
+                Ok(0)
+            }
+        }
+    }
+
+    #[cfg(test)]
+    #[test]
+    fn load_files() {
+        let loader = LocalFsReader {
```

Comment:
This test fails since I added randomness to the split algorithm (returns 3 some times).